task/import.go

```diff
@@ -0,0 +1,131 @@
+package task
+
+import (
+	"bytes"
+	"context"
+	"encoding/json"
+	"fmt"
+	"io"
+	"mime"
+	"net/http"
+	"path"
+	"time"
+
+	livepeerAPI "github.com/livepeer/go-api-client"
+	"github.com/livepeer/go-livepeer/drivers"
+	"github.com/livepeer/livepeer-data/pkg/data"
+	"golang.org/x/sync/errgroup"
+)
+
+const (
+	fileUploadTimeout = 5 * time.Minute
+	videoFileName     = "video"
+	metadataFileName  = "video.json"
+)
+
+func TaskImport(tctx *TaskContext) (*data.TaskOutput, error) {
+	var (
+		ctx        = tctx.Context
+		playbackID = tctx.Asset.PlaybackID
+		params     = *tctx.Task.Params.Import
+		osSess     = tctx.osSession
+	)
+	filename, contents, err := getFile(ctx, osSess, params)
+	if err != nil {
+		return nil, err
+	}
+	defer contents.Close()
+
+	secondaryReader, pipe := io.Pipe()
+	mainReader := io.TeeReader(contents, pipe)
+
+	eg, egCtx := errgroup.WithContext(ctx)
+	var (
+		videoFilePath, metadataFilePath string
+		metadata                        *FileMetadata
+	)
+	eg.Go(func() (err error) {
+		metadata, err = Probe(egCtx, filename, mainReader)
+		pipe.CloseWithError(err)
+		if err != nil {
+			return err
+		}
+		metadataFilePath, err = saveMetadataFile(egCtx, osSess, playbackID, metadata)
+		return err
+	})
+	eg.Go(func() (err error) {
+		fullPath := path.Join(playbackID, videoFileName)
+		videoFilePath, err = osSess.SaveData(egCtx, fullPath, secondaryReader, nil, fileUploadTimeout)
+		if err != nil {
+			return fmt.Errorf("error uploading file=%q to object store: %w", videoFileName, err)
+		}
+		return nil
+	})
+	if err := eg.Wait(); err != nil {
+		// TODO: Delete the uploaded file
+		return nil, err
+	}
+	return &data.TaskOutput{Import: &data.ImportTaskOutput{
+		VideoFilePath:    videoFilePath,
+		MetadataFilePath: metadataFilePath,
+		AssetSpec:        metadata.AssetSpec,
+	}}, nil
+}
+
+func getFile(ctx context.Context, osSess drivers.OSSession, params livepeerAPI.ImportTaskParams) (string, io.ReadCloser, error) {
+	if upedObjKey := params.UploadedObjectKey; upedObjKey != "" {
+		// TODO: We should simply "move" the file in case of direct import since we
+		// know the file is already in the object store. Independently, we also have
+		// to delete the uploaded file after copying to the new location.
+		fileInfo, err := osSess.ReadData(ctx, upedObjKey)
+		if err != nil {
+			return "", nil, UnretriableError{fmt.Errorf("error reading direct uploaded file: %w", err)}
+		}
+		return fileInfo.FileInfo.Name, fileInfo.Body, nil
+	} else if params.URL == "" {
+		return "", nil, fmt.Errorf("no import URL or direct upload object key: %+v", params)
+	}
+
+	req, err := http.NewRequestWithContext(ctx, "GET", params.URL, nil)
+	if err != nil {
+		return "", nil, UnretriableError{fmt.Errorf("error creating http request: %w", err)}
+	}
+	resp, err := http.DefaultClient.Do(req)
+	if err != nil {
+		return "", nil, fmt.Errorf("error on import request: %w", err)
+	}
+	if resp.StatusCode >= 300 {
```

Comment:
acc. to the `CheckRedirect` setting in [http.Client](https://pkg.go.dev/net/http#Client), it appears that redirects are followed by default. you can ignore the above comment (however >= 400 status signifies errors)

Reply:
This conceptually means an "if not a 2xx". I don't think we need to follow redirects since users should create the import task with the URL that returns the file, not one that will redirect to something else. But I don't think we need to go out of our way to avoid following redirects anyway so we can keep the default behavior. And it seems like that default behavior is to follow up to 10 redirects on a request before just returning the `3xx` response. On that case, I do agree that we should treat it as an error, and also an error we should not retry just like a `4xx`. If the user wants to try again sometime in the future after fixing their server redirection behavior, they can create another task.